pkg/controllers/user/globaldns/user_cluster_globaldns_controller.go

```diff
@@ -56,67 +77,130 @@ func (g *UserGlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.O
 		return nil, err
 	}
 
-	// go through target projects which are part of the current cluster
+	// go through target projects which are part of the current cluster and find all ingresses
+	var allIngresses []*v1beta1.Ingress
+
 	for _, t := range mcapp.Spec.Targets {
 		split := strings.SplitN(t.ProjectName, ":", 2)
 		if len(split) != 2 {
-			return mcapp, fmt.Errorf("error in splitting project ID %v", t.ProjectName)
+			return nil, fmt.Errorf("error in splitting project ID %v", t.ProjectName)
 		}
 		// check if the target project in this iteration is same as the cluster in current context
 		if split[0] != g.clusterName {
 			logrus.Debugf("Continuing since target is not for current cluster since project %v doesn't belong in cluster %v", split[1], g.clusterName)
 			continue
 		}
 
-		updated, err := g.updateGlobalDNSEndpointsForTarget(&t, obj)
+		// each target will have appName, this appName is also the namespace in which all workloads for this app are created
+		ingresses, err := g.ingressLister.List(t.AppName, labels.NewSelector())
 		if err != nil {
-			return updated, err
+			return nil, err
 		}
+		allIngresses = append(allIngresses, ingresses...)
 	}
-	return nil, err
+
+	//gather endpoints
+	return g.fetchGlobalDNSEndpointsForIngresses(allIngresses, obj)
 }
 
-func (g *UserGlobalDNSController) updateGlobalDNSEndpointsForTarget(t *v3.Target, obj *v3.GlobalDNS) (*v3.GlobalDNS, error) {
-	// each target will have appName, this appName is also the namespace in which all workloads for this app are created
-	ingresses, err := g.ingressLister.List(t.AppName, labels.NewSelector())
-	if err != nil {
-		return nil, err
+func (g *UserGlobalDNSController) reconcileProjects(obj *v3.GlobalDNS) ([]string, error) {
+	// go through target projects which are part of the current cluster and find all ingresses
+	var allIngresses []*v1beta1.Ingress
+
+	for _, projectNameSet := range obj.Spec.ProjectNames {
+		split := strings.SplitN(projectNameSet, ":", 2)
+		if len(split) != 2 {
+			return nil, fmt.Errorf("UserGlobalDNSController: Error in splitting project Name %v", projectNameSet)
+		}
+		// check if the project in this iteration belongs to the same cluster in current context
+		if split[0] != g.clusterName {
+			logrus.Debugf("UserGlobalDNSController: Continuing since project %v doesn't belong in cluster %v", split[1], g.clusterName)
+			continue
+		}
+		projectID := split[1]
+		//list all namespaces in this project and list all ingresses within each namespace
+		namespaces, err := g.namespaceLister.List("", labels.NewSelector())
```

Comment:
we should either list namespaces outside of the list project loop (as we don't want to iterate over the same set all over again on every iteration). Or we can use the indexer to fetch the namespaces for a specific project. Here is the example: https://github.com/rancher/rancher/blob/master/pkg/controllers/user/resourcequota/resource_quota_calculate_used.go#L56

Reply:
done, listing out of the loop